File: cpp/98. Validate Binary Search Tree.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
// Inorder of BST is definitely sorted
public:
    bool dfs(TreeNode *root, int &pre, bool &isFirst) {
        if(!root) return true;
        if(!dfs(root->left, pre, isFirst)) return false;
        if(!isFirst && root->val <= pre) return false;
        pre = root->val;
        isFirst = false;
        if(!dfs(root->right, pre, isFirst)) return false;
        return true;
    }
    bool isValidBST(TreeNode* root) {
        int pre = INT_MIN;
        bool isFirst = true;
        return dfs(root, pre, isFirst);
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}

File: python/98-Validate Binary Search Tree.py
from bintree_helper import *

'''
"ValidBST <=> Inorder is sorted
'''

class Solution:
    def isValidBST(self, root) -> bool:
        "Recursive In-Order Traversal, Time: O(n), Space: O(n)"
        prev = float("-inf")
        def dfs(root):
            nonlocal prev
            if root.left: 
                if dfs(root.left) == False: return False
            if root.val <= prev: return False
            prev = root.val
            if root.right: 
                if dfs(root.right) == False: return False
            return True
        return dfs(root)
    def isValidBST(self, root) -> bool:
        "Iterative In-Order Traversal, Time: O(n), Space: O(n)"
        from collections import deque
        stack = deque()
        def helper(root):
            while root:
                stack.append(root)
                root = root.left
        helper(root)
        prev = float("-inf")
        while stack:
            node = stack.pop()
            if node.val <= prev: return False
            prev = node.val
            if node.right:
                helper(node.right)
        return True
root = deserialize("[2,1,3]")
result = Solution().isValidBST(root)
assert result == True

root = deserialize("[5,1,4,null,null,3,6]")
result = Solution().isValidBST(root)
assert result == False

root = deserialize("[1,1]")
result = Solution().isValidBST(root)
assert result == False